src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -507,6 +516,76 @@ export class SearchSource {
     return searchRequest;
   }
 
+  private flattenWithFieldsApi(searchRequest: Record<string, any>) {
+    const { getConfig } = this.dependencies;
+
+    searchRequest.body = searchRequest.body || {};
+    const { body, index, query, filters, highlightAll } = searchRequest;
+    searchRequest.indexType = this.getIndexType(index);
+
+    const fieldListProvided = !!body.fields;
+
+    // get some special field types from the index pattern
+    const { storedFields, scriptFields, docvalueFields } = index
+      ? index.getComputedFields()
+      : {
+          storedFields: ['*'],
+          scriptFields: {},
+          docvalueFields: [],
+        };
+
+    // add docvalues to fields if no fields are explicitly provided
+    body.fields = fieldListProvided ? body.fields : docvalueFields;
+
+    // apply source filters from index pattern if specified by the user
+    if (index) {
+      const sourceFilters = index.getSourceFiltering();
+      if (!body.hasOwnProperty('_source')) {
+        body._source = sourceFilters;
+      }
+      const filter = fieldWildcardFilter(
+        sourceFilters.excludes,
+        getConfig(UI_SETTINGS.META_FIELDS)
+      );
+      body.fields = body.fields.filter((fld: any) =>
+        filter(typeof fld === 'string' ? fld : fld.field)
+      );
+    }
+
+    // add scripted fields
+    body.script_fields = pick(Object.assign(body.script_fields || {}, scriptFields), body.fields);
+    body.fields = body.fields.filter((fld: any) => {
+      const f = typeof fld === 'string' ? fld : fld.field;
+      return !Object.keys(body.script_fields).includes(f);
+    });
+
+    // request the remaining fields from stored_fields just in case, since the
+    // fields API does not handle stored fields
+    if (fieldListProvided) {
+      const bodyFieldNames = body.fields.map((field: any) =>
+        typeof field === 'string' ? field : field.field
+      );
+      const remainingFields = difference(bodyFieldNames, keys(body.script_fields));
+      body.stored_fields = remainingFields;
+
+      // remove _source, since everything's coming from fields API, scripted fields,
+      // or stored fields
+      body._source = false;
+    } else {
+      body.stored_fields = storedFields;
```

Comment:
why are we still fetching from _source in this case ?

Reply:
This was one of my questions above (Should we exclude _source when no fields are explicitly provided and instead request all of the fields which are available on the index pattern?)

If we add a `fieldsFromSource` option as you suggested, then I think we could always exclude _source unless particular fields are requested for it.